src/S3/S3Transfer/MultipartDownloader.php

```diff
@@ -0,0 +1,497 @@
+                            $this->partDownloadCompleted($result, $this->currentPartNo);
+
+                            return $result;
+                        })->otherwise(function ($reason) use ($nextCommand) {
+                            $this->partDownloadFailed($nextCommand, $reason, $this->currentPartNo);
+
+                            return $reason;
+                        });
+                } catch (\Throwable $e) {
+                    $this->downloadFailed($e, $this->objectCompletedPartsCount, $this->objectBytesTransferred, $this->currentPartNo);
+                    // TODO: yield transfer exception modeled with a transfer failed response.
+                    yield Create::rejectionFor($e);
+                }
+
+            }
+
+            // Transfer completed
+            $this->objectDownloadCompleted();
+
+            // TODO: yield the stream wrapped in a modeled transfer success response.
+            yield Create::promiseFor(new DownloadResponse(
+                $this->stream,
+                []
+            ));
+        });
+    }
+
+    /**
+     * Returns the next command for fetching the next object part.
+     *
+     * @return CommandInterface
+     */
+    abstract protected function nextCommand() : CommandInterface;
+
+    /**
+     * Compute the object dimensions, such as size and parts count.
+     *
+     * @param ResultInterface $result
+     *
+     * @return void
+     */
+    abstract protected function computeObjectDimensions(ResultInterface $result): void;
+
+    /**
+     * Calculates the object size dynamically.
+     *
+     * @param $sizeSource
+     *
+     * @return int
+     */
+    protected function computeObjectSize($sizeSource): int
+    {
+        if (is_int($sizeSource)) {
+            return (int) $sizeSource;
+        }
+
+        if (empty($sizeSource)) {
+            throw new \RuntimeException('Range must not be empty');
+        }
+
+        // For extracting the object size from the ContentRange header value.
+        if (preg_match("/\/(\d+)$/", $sizeSource, $matches)) {
+            return $matches[1];
+        }
+
+        throw new \RuntimeException('Invalid source size format');
+    }
+
+    /**
+     * MultipartDownloader factory method to return an instance
+     * of MultipartDownloader based on the multipart download type.
+     *
+     * @param S3ClientInterface $s3Client
+     * @param string $multipartDownloadType
+     * @param array $requestArgs
+     * @param array $config
+     * @param int $currentPartNo
+     * @param int $objectPartsCount
+     * @param int $objectCompletedPartsCount
+     * @param int $objectSizeInBytes
+     * @param int $objectBytesTransferred
+     * @param string $eTag
+     * @param string $objectKey
+     * @param MultipartDownloadListener|null $listener
+     * @param TransferListener|null $progressTracker
+     *
+     * @return MultipartDownloader
+     */
+    public static function chooseDownloader(
```

Comment:
This has too many parameters. But I think having so many parameters for each of the upload/download classes is the root cause-  Would it be possible to use a more consolidated config array (or possibly an object) to alleviate this?